data/finset.lean

```diff
@@ -845,4 +845,25 @@ sort_eq _ _
 list.to_finset_eq (sort_nodup r s) ▸ eq_of_veq (sort_eq r s)
 end sort
 
+/- not mem -/
+section not_mem
+variables {s s₁ s₂ : finset α} [decidable_eq α] {a b : α}
+
+theorem ne_of_mem_of_not_mem (h_mem : a ∈ s) (h_not_mem : b ∉ s) : a ≠ b :=
+ne_of_mem_erase (@@eq.substr _ (erase_insert h_not_mem) h_mem)
```

Comment:
You know, this can be proved for literally anything that has a `has_mem`, it has nothing to do with finsets.

Reply:
Fair enough. For whatever reason that didn't occur to me before.

```lean
theorem ne_of_mem_of_not_mem {A S : Type} [has_mem A S] {a b : A} {s : S} (h_mem : a ∈ s) (h_not_mem : b ∉ s) : a ≠ b :=
λ p : a = b, by subst p; contradiction
```

Reply:
Now that I think about it, does this not already exist somewhere?

```lean
theorem ne_of_p_of_not_p {A : Type} {a b : A} (P : A → Prop) (pa : P a) (npb : ¬ P b) : a ≠ b :=
λ h, by subst h; contradiction

theorem ne_of_mem_of_not_mem {A S : Type} [has_mem A S] {a b : A} {s : S} : a ∈ s → b ∉ s → a ≠ b :=
ne_of_p_of_not_p (λ a : A, a ∈ s)
```

Reply:
Let me know how you'd like the PR changed. I can add the above two theorems, though they would not naturally go into `data/finset.lean`.